lasagne/layers/autoenc.py

```diff
@@ -0,0 +1,262 @@
+    respect to its input: transposed layer
+    for a :class:`DenseLayer`, deconvolutional layer for
+    :class:`Conv2DLayer`, :class:`Conv1DLayer`; or
+    an unpooling layer for :class:`MaxPool2DLayer`.
+
+    It is specially useful for building (convolutional)
+    autoencoders with tied parameters.
+
+    Note that if the layer to be inverted contains a nonlinearity
+    and/or a bias, the :class:`InverseLayer` will include the derivative
+    of that in its computation.
+
+    Parameters
+    ----------
+    incoming : a :class:`Layer` instance or a tuple
+        The layer feeding into this layer, or the expected input shape.
+    layer : a :class:`Layer` instance or a tuple
+        The layer with respect to which the instance of the
+        :class:`InverseLayer` is inverse to.
+
+    Examples
+    --------
+    >>> import lasagne
+    >>> from lasagne.layers import InputLayer, Conv2DLayer, DenseLayer
+    >>> from lasagne.layers import InverseLayer
+    >>> l_in = InputLayer((100, 3, 28, 28))
+    >>> l1 = Conv2DLayer(l_in, num_filters=16, filter_size=5)
+    >>> l2 = DenseLayer(l1, num_units=20)
+    >>> l_u = InverseLayer(l2, l1)  # As Deconv2DLayer
+    """
+    def __init__(self, incoming, layer, **kwargs):
+
+        super(InverseLayer, self).__init__(
+            [incoming, layer, layer.input_layer], **kwargs)
+
+    def get_output_shape_for(self, input_shapes):
+        return input_shapes[2]
+
+    def get_output_for(self, inputs, **kwargs):
+        input, layer_out, layer_in = inputs
+        return theano.grad(None, wrt=layer_in, known_grads={layer_out: input})
+
+
+def build_autoencoder(layer, nonlinearity='same', b='new'):
+    """
+    Given a :class:`Layer` instance, this function builds a
+    symmetric autoencoder with tied weights.
+
+    Parameters
+    ----------
+    layer: a :class:`Layer` instance
+        The :class:`Layer` instance with respect to which a symmetric
+        autoencoder is built.
+    nonlinearity : 'same', list, callable, or None
+        The nonlinearity that is applied to the layer activations of theano
+        :class:`InverseLayer`. If None is provided, the layer will be linear.
+        If 'same', the :class:`InverseLayer` has the same nonlinearity as the
+        mirror :class:`Layer`.
+    b : 'new', list, Theano shared variable, numpy array, callable or None
+        An initializer for the biases. If 'new', new biases are initialized
+        using the default initialization. If a shared variable or a numpy array
+        is provided, the shape must match the incoming shape, skipping those
+        axes the biases are shared over (see below for an example). If set to
+        ``None``, the layer will have no biases and pass through its input
+        unchanged.
+        See :func:`lasagne.utils.create_param` for more information.
+
+    Returns
+    -------
+    layer: :class:`Layer` instance
+       The output :class:`Layer` of the symmetric autoencoder with
+       tied weights.
```

Comment:
It also has to return the code layer of the autoencoder. As you're changing the encoder in `unfold_bias_and_nonlinearity_layers()`, the `layer` the user put into the function is not the code layer any more, because it has got its bias and nonlinearity stripped off. You need to return the code layer and explain in the documentation that the `layer` given by the user is not valid any more.